src/epanet.c

```diff
@@ -3166,6 +3302,352 @@ int DLLEXPORT ENgetaveragepatternvalue(int index, EN_API_FLOAT_TYPE *value)
+    Tank[Ntanks].Vcurve = 0;
+    Tank[Ntanks].MixModel = 0;
+    Tank[Ntanks].V1max = 10000;
+  }
+  
+  Nnodes++;
+
+  /* set default values for new node */
+  strncpy(Node[n].ID, id, MAXID);
+  
+  
+  Node[n].El = 0;
+  Node[n].S = NULL;
+  Node[n].C0 = 0;
+  Node[n].Ke = 0;
+  Node[n].Rpt = 0;
+  Coord[n].HaveCoords = FALSE;
+  Coord[n].X = 0;
+  Coord[n].Y = 0;
+  
+  /* Insert new node into hash table */
+  ENHashTableInsert(NodeHashTable, Node[n].ID, n);        /* see HASH.C */
+  return(0);
+}
+
+int DLLEXPORT ENaddlink(char *id, EN_LinkType linkType, char *fromNode, char *toNode)
+{
+  int i, n;
+  
+/* Check if a link with same id already exists */
+  if ( !Openflag ) return(102);
+  if ( ENgetlinkindex(id, &i) == 0 ) return(215);
+  
+/* Lookup the from and to nodes */
+  int N1 = ENHashTableFind(NodeHashTable, fromNode);
+  int N2 = ENHashTableFind(NodeHashTable, toNode);
+  
+  if (N1 == 0 || N2 == 0) {
+    return(203);
+  }
+  
+/* Check that id name is not too long */
+  if (strlen(id) > MAXID) return(250);
+  
+  Nlinks++;
+  n = Nlinks;
+  
+/* Grow arrays to accomodate the new value */
+  Link = (Slink *)realloc(Link, (Nlinks+1)*sizeof(Slink));
+  Q = (double *)realloc(Q, (Nlinks+1)*sizeof(double));
+  LinkSetting = (double *)realloc(LinkSetting, (Nlinks+1)*sizeof(double));
+  LinkStatus = (char *)realloc(LinkStatus, (Nlinks+1)*sizeof(char));
+  
+  strncpy(Link[n].ID, id, MAXID);
+  
+  if (linkType <= EN_PIPE) {
+    Npipes++;
+  }
+  else if (linkType == EN_PUMP) {
+    Npumps++;
+    
+    /* Grow pump array to accomodate the new value */
+    Pump = (Spump *)realloc(Pump, (Npumps+1)*sizeof(Spump));
+    
+    Pump[Npumps].Link = n;
+    Pump[Npumps].Ptype = 0;
+    Pump[Npumps].Q0 = 0;
+    Pump[Npumps].Qmax = 0;
+    Pump[Npumps].Hmax = 0;
+    Pump[Npumps].H0 = 0;
+    Pump[Npumps].R = 0;
+    Pump[Npumps].N = 0;
+    Pump[Npumps].Hcurve = 0;
+    Pump[Npumps].Ecurve = 0;
+    Pump[Npumps].Upat = 0;
+    Pump[Npumps].Epat = 0;
+    Pump[Npumps].Ecost = 0;
+    Pump[Npumps].Energy[5] = MISSING;
+    
+  }
+  else {
+    
+    /* Grow valve array to accomodate the new value */
+    Nvalves++;
+    Valve = (Svalve *)realloc(Valve, (Nvalves+1)*sizeof(Svalve));
+    Valve[Nvalves].Link = n;
+  }
+
+  Link[n].Type = linkType;
+  Link[n].N1 = N1;
+  Link[n].N2 = N2;
+  
+  if(linkType == PUMP) {
+    Link[n].Diam = Npumps;
+  } else {
+    Link[n].Diam = 0;
+  }
+  Link[n].Len = 0;
+  Link[n].Kc  = 0.01;
+  Link[n].Km  = 0;
+  Link[n].Kb  = 0;
+  Link[n].Kw  = 0;
+  Link[n].R  = 0;
+  Link[n].Rc  = 0;
+  Link[n].Stat = 0;
+  Link[n].Rpt = 0;
+  
+  ENHashTableInsert(LinkHashTable, Link[n].ID, n);
+  return(0);
+}
+
+int ENdeletelink(int index)
+{
+  int i, linkType;
+  
+  if (!Openflag) return(102);
+  if (index <= 0 || index > Nlinks) return(203);
+  
+  ENgetlinktype(index, &linkType);
+  
+  // remove from hash table
+  ENHashTableDelete(LinkHashTable, Link[index].ID);
+  ENHashTableFind(LinkHashTable, Link[index].ID);

```

Comment:
probably want to check for error on delete, then ensure that `find` returns `NOTFOUND` ?